enhancements/ingress/logging-headers.md

```diff
@@ -0,0 +1,260 @@
+configure capture of HTTP request and response headers in access logs.  By
+default, when access logging is enabled, the IngressController logs the HTTP
+request line and response code but no headers.  Using the new API, users may
+specify a list of arbitrary HTTP request headers and a list of response headers
+that should be captured and logged when access logging is enabled.
+
+## Motivation
+
+Certain headers may need to be logged for compliance reasons or other reasons.
+In OpenShift 3, cluster administrators could use a custom HAProxy configuration
+template to customize logging, but using a custom template is not permitted in
+OpenShift 4.
+
+### Goals
+
+1. Enable the cluster administrator to specify HTTP request and response headers that should be captured.
+2. Enable the cluster administrator to specify a custom HTTP log format string that references the captured headers.
+
+### Non-Goal
+
+1. Providing control on a per-frontend or per-route basis.
+
+## Proposal
+
+The IngressController API is extended by adding an optional `HTTPCaptureHeaders`
+field with type `*IngressControllerCaptureHTTPHeaders` to `AccessLogging`:
+
+```go
+type AccessLogging struct {
+	// ...
+
+	// httpCaptureHeaders defines HTTP headers that should be captured in
+	// access logs.  If this field is empty, no headers are captured.
+	//
+	// +optional
+	HTTPCaptureHeaders *IngressControllerCaptureHTTPHeaders `json:"httpCaptureHeaders,omitempty"`
+}
+```
+
+The `IngressControllerCaptureHTTPHeaders` type has a `Request` field and a
+`Response` field, both of type `[]IngressControllerCaptureHTTPHeader`, for
+specifying headers to capture:
+
+```go
+// IngressControllerCaptureHTTPHeaders specifies which HTTP headers the
+// IngressController captures.
+type IngressControllerCaptureHTTPHeaders struct {
+	// request specifies which HTTP request headers to capture.
+	//
+	// If this field is empty, no request headers are captured.
+	Request []IngressControllerCaptureHTTPHeader `json:"request,omitempty"`
+
+	// response specifies which HTTP response headers to capture.
+	//
+	// If this field is empty, no response headers are captured.
+	Response []IngressControllerCaptureHTTPHeader `json:"response,omitempty"`
+}
+```
+
+The `IngressControllerCaptureHTTPHeader` type has fields for specifying a header
+name and maximum length for a captured header value (header values that exceed
+the maximum length are truncated to that length):
+
+```go
+// IngressControllerCaptureHTTPHeader describes an HTTP header that should be
+// captured.
+type IngressControllerCaptureHTTPHeader struct {
+	// name specifies a header name.  Its value must be a valid HTTP header
+	// name.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^[\x21\x23-\x27\x2a\x2b\x2d\x2e\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+$`
+	// +required
+	Name string `json:"name"`
+
+	// maxLength specifies a maximum length for the header value.  If a
```

Comment:
Is the length measured in UTF-8 characters?

Reply:
Seems to be raw bytes:  http://git.haproxy.org/?p=haproxy-2.0.git;a=blob;f=src/sample.c;h=5bd46f8caa6c6b3ab6b404b201e144535dcfee48;hb=HEAD#l671